### DISCUSSION TITLE
How can we better support a wider range of reactors?

### DISCUSSION THREAD

**Discussion by john-science** [terrapower/armi#1894] [Ideas]
I believe the ARMI framework was originally too focused on pin-type reactors with hex assemblies, so that will go on the list. But where at the other frustration points for non-Natrium users/developers?

* [ ] [makes too many hex-assembly assumptions in the Core](https://github.com/terrapower/armi/issues/1448)
* [ ] [does not support ex-core structures](https://github.com/terrapower/armi/issues/262)
* [ ] enforces "Axial Expansion Changer", even if the reactor does not have assemblies
* [ ] assumes people want to do any thermal expansion at all
* [ ] assumes people will be using a lattice physics code
* [ ] assumes users will want a burn chain defined in the framework
* [ ] assumes nuclide flags are useful for everyone

**Comment by john-science** [comment]
@drewj-tp @keckler @sombrereau @alexhjames @jeffbaylor What do YOU guys think? Help me build this list!

**Comment by drewj-tp** [comment]
cc @slee-usnc @dturkoglu

**Reply by john-science** [reply]
Hey @drewj-tp , it looks to me like a couple of your points above could be solved by just tweaking your `App`.  Maybe I could build a tutorial example for this. But let's you do: `class MyApp(App)`, you are subclassing `App` and so you get a default set of plugins:

https://github.com/terrapower/armi/blob/f3d16df8a1946039c0a60b0650ced60fb0600a7a/armi/apps.py#L87

https://github.com/terrapower/armi/blob/f3d16df8a1946039c0a60b0650ced60fb0600a7a/armi/apps.py#L98-L106

Well, when in your `MyApp` class you could over-write the default `__initNewPlugins()` method and just comment out the `ThermalHydraulicsPlugin` and maybe even the `NeutronicsPlugin`.

I'm just thinking aloud, but if some of your problems are that ARMI forces certain functionality through the default plugins in the default `App`, it is meant to be very easy to remove plugins from the default App.

Thoughts?

**Reply by drewj-tp** [reply]
That could be useful. I personally get squeamish overwriting methods with leading double underscores.

It doesn't really eliminate the problem in my other comment about [`Block` does too much SFR specific stuff](https://github.com/terrapower/armi/discussions/1894#discussioncomment-10697345) because all block-like things still have to inherit from `Block`. But it is a way to help [ARMI do less by default](https://github.com/terrapower/armi/discussions/1896)

**Reply by john-science** [reply]
> I personally get squeamish overwriting methods with leading double underscores.

Let's not worry about that right now. If it solves a problem, I can remove the double-underscore or change the system at a higher-level to make this a part of the ARMI API.

> It doesn't really eliminate the problem in my other comment about Block

Sure, sure. I wasn't suggesting it would solve ALL your issues. But this seems like an EASY solution to turn off: Axial Expansion, thermal expansion, and lattice physics. Or so it seems to me.

**Comment by drewj-tp** [comment]
Some cherry-picked examples

- `Block.__init__` sets [`bondRemoved`, `displacementX` and `displacementY`](https://github.com/terrapower/armi/blob/f3d16df8a1946039c0a60b0650ced60fb0600a7a/armi/reactor/blocks.py#L114-L119) to zero which mean a lot to SFRs but not to things that aren't SFRs
- `Block.p.puFrac` being set at [block construction](https://github.com/terrapower/armi/blob/f3d16df8a1946039c0a60b0650ced60fb0600a7a/armi/reactor/blocks.py#L784-L786) makes a lot of sense for some fast reactors, mox-fueled reactors, but not as much (at construction) for LWRs
  - This is more generally useful through depletion to look at plutonium buildup. But not always the case.
- [`Block.p.smearDensity`](https://github.com/terrapower/armi/blob/f3d16df8a1946039c0a60b0650ced60fb0600a7a/armi/reactor/blocks.py#L788-L792) doesn't make sense for non-pinned reactors
- [`Block.getNumPins`](https://github.com/terrapower/armi/blob/f3d16df8a1946039c0a60b0650ced60fb0600a7a/armi/reactor/blocks.py#L1047) is a thing that makes sense for pin-type reactors
  - But there's also [`Block.p.nPins`](https://github.com/terrapower/armi/blob/f3d16df8a1946039c0a60b0650ced60fb0600a7a/armi/reactor/blockParameters.py#L774)
- [`HexBlock.verifyBlockDims` assumes you have wire, duct, and clad](https://github.com/terrapower/armi/blob/f3d16df8a1946039c0a60b0650ced60fb0600a7a/armi/reactor/blocks.py#L2170-L2175) components or else it [yells at you](https://github.com/terrapower/armi/blob/f3d16df8a1946039c0a60b0650ced60fb0600a7a/armi/reactor/blocks.py#L2178-L2181).
- [`HexBlock` get/has pin pitch](https://github.com/terrapower/armi/blob/f3d16df8a1946039c0a60b0650ced60fb0600a7a/armi/reactor/blocks.py#L2400-L2406) [assumes you have clad and wire or else your block is not well defined](https://github.com/terrapower/armi/blob/f3d16df8a1946039c0a60b0650ced60fb0600a7a/armi/reactor/blocks.py#L2432-L2439)

**Reply by john-science** [reply]
Yeah, a better organization might have been (just for example): 

* `Composite`
    * `Block`
        * `PinnedBlock`
            * `PinnedHexBlock`
            * `PinnedSquareBlock`

More layers of abstraction would give us more control over the methods and parameters on these objects. I can agree to that.

It would also, bonus, make our classes and modules much shorter and easier to parse.

**Comment by alexhjames** [comment]
I don't have many things with clear things to change but here's what I can think of:

- Philosophically, ARMI is very geared around running ARMI with physics plugins to generate results. Where possible, it would be good to embrace ARMI's additional role as a data model that people look at to get the results of the physics analyses. ARMI and it's database is a very convenient form to pass around at analysis interfaces to communicate data. For example, some of the problems with loading multiple reactors. From a data model aspect, this is a reasonable ask because you need to look at multiple things for data through time, but isn't as relevant if you're just thinking in terms of running an ARMI simulation for a single reactor. 
- Axial expansion assumes free expansion which isn't necessarily true depending on assembly design. 
- ARMI assumes that all physics will be run at a single mesh snapshot and then things are updated, then things are run again at the next mesh snapshot. This allows for things like remeshing between time nodes, or having assemblies appear in the core during a time node in the middle of the analysis. This is not true for other physics codes which need to simulate through time and therefore have a static mesh that is set at the beginning of time. There isn't necessarily something to change right now but maybe there could be other helper methods that say interpolate results back on the BOL mesh elevations, or that collect all assemblies that will every be in the core at any point in the analysis so that they can be meshed at BOL (and put in like a load queue until they're needed). 
- ARMI does poorly with spatially varying quantities. It is common to find errors every so often related to a spatial varying quantity not changing when it should have (like grow to full core, or shuffling, etc). There are lots of assumptions about smearing things over the block/assembly which are not valid for all parameters.
